### PR TITLE
HPCC-18517 Remove ESP direct calls to validateCache() and clean

### PR DESCRIPTION
**Pull request by wangkx**
1. add a default update=false param into getEnvironmentFactory(); If
   the update=true, call validateCache();
2. in ESP services, remove ESP direct calls to IEnvironmentFactory::
   validateCache(); also change getEnvironmentFactory() calls to
   getEnvironmentFactory(true);
3. add more error checks; Remove old code (not used):
   getClusterConfig() and CWsDfuEx::getRoxieClusterConfig().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Start an EclWatch. Click Operation icon. Select System Servers tab. Check whether all of the system servers are displayed or not.

### CODE REVIEW

**Comment by wangkx** [comment]
@jakesmith please review.

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 1]
```diff
@@ -214,9 +214,9 @@ bool CWsEclService::init(const char * name, const char * type, IPropertyTree * c
             daliAddress.append(*daliServers++);
     }
 
-    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
-    Owned<IConstEnvironment> environment = factory->openEnvironment();
-    Owned<IPropertyTree> pRoot = &environment->getPTree();
+    Owned<IPropertyTree> pRoot = getEnvironmentPTreeWithUpdate();
+    if (!pRoot)
+        throw MakeStringException(-1, "Failed to get environment information.");
```
Comment (jakesmith) [inline esp/services/ws_ecl/ws_ecl_service.cpp]: See comment above about accessing raw meta info.
I think this is a different change from the issue raised in the JIRA + should be tackled not by exposing the meta info more directly, but by extending the IConstEnvironment interfaces where appropriate to access the info. you need.

For example here, I think you're walking RoxieCluster's .. if there isn't already methods to give you a list of Roxie's and/or not a good way to get to their instance addresses, the environment interfaces should be extended.

Should open a separate JIRA for that though.

File: esp/services/ws_fs/ws_fsBinding.cpp [hunk 2]
```diff
@@ -200,9 +200,11 @@ int CFileSpraySoapBindingEx::onGetInstantQuery(IEspContext &context, CHttpReques
...
 IPropertyTree* CFileSpraySoapBindingEx::createPTreeForXslt(double clientVersion, const char* method, const char* dfuwuid)
 {
-    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
-    Owned<IConstEnvironment> m_constEnv = factory->openEnvironment();
-    Owned<IPropertyTree> pEnvRoot = &m_constEnv->getPTree();
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactoryWithUpdate();
+    Owned<IConstEnvironment> constEnv = factory->openEnvironment();
+    if (!constEnv)
+        throw MakeStringException(ECLWATCH_CANNOT_GET_ENV_INFO, "Failed to get environment information.");
+    Owned<IPropertyTree> pEnvRoot = &constEnv->getPTree();
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsBinding.cpp]: should really be trying to avoid this, see other comments (but it's separate issue)

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@wangkx - see comments.

I think you should not add the new getEnvironmentPTreeWithUpdate() here and open a separate JIRA for the changes that involve code that is directly walking the environment tree at the moment.

**Comment by wangkx** [comment]
@jakesmith I am wondering which way is more efficient if I need to get >1 groups of environment settings in one ESP method: retrieving/parsing the raw meta info or calling >1 environment interface methods (repeatedly getEnvironmentFactory() and openEnvironment()). Please advise.

**Comment by jakesmith** [comment]
@wangkx - >1 IConstEnvironment method call is fine/the way to go, it shouldn't be any more costly (or very little) to call a couple of methods to get the info. back, then for you to walk the IPT manually.

For this PR though, you only need the validateCache change, other changes relating to cleaning up/encapsulating the requests in environment.cpp can wait for another JIRA/PR.

**Comment by wangkx** [comment]
Ok, I will do...

**Comment by wangkx** [comment]
@jakesmith please review. The new JIRA for encapsulating the requests in environment.cpp: https://track.hpccsystems.com/browse/HPCC-18654

**Comment by jakesmith** [comment]
Looks fine.
Should add something to the "Testing:" section of the PR is empty though.
@wangkx

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

File: common/environment/environment.cpp [hunk 3]
```diff
@@ -1829,14 +1829,16 @@ unsigned CConstDropZoneInfoIterator::count() const
...
-extern ENVIRONMENT_API IEnvironmentFactory * getEnvironmentFactory()
+extern ENVIRONMENT_API IEnvironmentFactory * getEnvironmentFactory(bool update)
 {
     CriticalBlock block(getEnvSect);
     if (!factory)
     {
         factory = new CEnvironmentFactory();
         addShutdownHook(*factory);
     }
+    if (update)
```
Comment (richardkchapman) [inline common/environment/environment.cpp]: Is this safe?

Any previous callers may still be holding on to their copies of factory, I think

Comment (wangkx) [reply common/environment/environment.cpp]: Good question. Should I move those 2 lines after line 1837?

Comment (richardkchapman) [reply common/environment/environment.cpp]: No.

I don't think it's a new issue - but if I am reading this code correctly the code in environment.cpp is working on the assumption that it is managing a singleton that is never updated, but that assumption was rendered invalid once calls to validateCache were added.

@jakesmith What do you think?

Comment (wangkx) [reply common/environment/environment.cpp]: They should be called after line 1839. By reading the code inside validateCache(), there are 2 functions: create the subscription and run handleEnvironmentChange. The handleEnvironmentChange calls clearCache(). Inside the clearCache(), the synchronized procedure(safeCache) should handle the thread safe, I think.

Comment (jakesmith) [reply common/environment/environment.cpp]: >it is managing a singleton that is never updated, but that assumption was rendered invalid once calls to validateCache were added.

There is only ever 1 factor (and 1 subscription).
The 1st call to validateCache will install the subscription. I though the idea then was that any notifications could clear the cache, so that subsequent calls to e.g. factory->openEnvironment() would reflect the update and not see old cached version.

However, handleEnvironmentChange() which resets the cache, is never automatically called. I would have expected it to have been called if the dirty flag had been set by the notifier on the next call to get the environment (e.g. next openEnvironment()).

Comment (jakesmith) [reply common/environment/environment.cpp]: @richardkchapman - I think it's thread safe, _if_ every interaction with the environment (via IConstEnvironment), locks out the cache clearing , including anything that's returned like e.g. a IConstDropZoneInfoIterator). 
And by the looks of it, that is the intent, every call is protected inside a 'synchronized procedure(safeCache)' that is locked when the cache is cleared.

**Comment by richardkchapman** [comment]
@ghalliday Can you check

File: common/environment/environment.hpp [hunk 4]
```diff
@@ -175,7 +175,7 @@ interface IEnvironmentFactory : extends IInterface
 
 
 class StringBuffer;
-extern "C" ENVIRONMENT_API IEnvironmentFactory * getEnvironmentFactory();
+extern "C" ENVIRONMENT_API IEnvironmentFactory * getEnvironmentFactory(bool update = false);
```
Comment (ghalliday) [inline common/environment/environment.hpp]: Would it be cleaner for this parameter to not have a default parameter - there are several cases which don't pass any parameter.  When should true be passed, and when should false?

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 5]
```diff
@@ -214,8 +214,10 @@ bool CWsEclService::init(const char * name, const char * type, IPropertyTree * c
             daliAddress.append(*daliServers++);
     }
 
-    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
     Owned<IConstEnvironment> environment = factory->openEnvironment();
```
Comment (ghalliday) [inline esp/services/ws_ecl/ws_ecl_service.cpp]: Should this be throwing an exception inside openEnvironment instead of having many places test it.  If there are valid situations where it is ok to continue if non-null then there could be another parameter to indicate that.

File: esp/services/WsDeploy/WsDeployEngine.cpp [hunk 6]
```diff
@@ -206,8 +206,7 @@ void CWsDeployEngine::deploy(CDeployOptions& pOptions)
 {
   try
   {
-    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
-    factory->validateCache();   
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
     Owned<IConstEnvironment> constEnv(factory->openEnvironment());
```
Comment (ghalliday) [inline esp/services/WsDeploy/WsDeployEngine.cpp]: See comment below.  Following code is invalid if it returns null

File: common/environment/environment.cpp (hunk 3 above)
Comment (ghalliday) [reply common/environment/environment.cpp]: I think it is probably safe as well.

**Comment by ghalliday** [comment]
@jakesmith what are your thoughts on my comments above.

File: esp/services/ws_ecl/ws_ecl_service.cpp (hunk 5 above)
Comment (jakesmith) [reply esp/services/ws_ecl/ws_ecl_service.cpp]: agree, in practice it's not possible for the openEnvironment implementation to return null anyway..

File: esp/services/WsDeploy/WsDeployEngine.cpp (hunk 6 above)
Comment (jakesmith) [reply esp/services/WsDeploy/WsDeployEngine.cpp]: agree, it should be assume must return non null, if openEnvironment could fail, it should throw an exception.

File: common/environment/environment.hpp (hunk 4 above)
Comment (jakesmith) [reply common/environment/environment.hpp]: Yes probably, always debatable on default params like this.

**Comment by jakesmith** [comment]
@wangkx - back to you.

File: esp/services/ws_ecl/ws_ecl_service.cpp (hunk 5 above)
Comment (wangkx) [reply esp/services/ws_ecl/ws_ecl_service.cpp]: @jakesmith I am not sure how to change the existing code to match with the new exception inside openEnvironment(). For example, in envGetConfigurationDirectory() (dalienv.cpp) called by queryDllServer(), if the openEnvironment() returns nullptr, a default value is used. If the openEnvironment() throws exception, should we catch the exception and use the default value?

Comment (wangkx) [reply esp/services/ws_ecl/ws_ecl_service.cpp]: Might we do need a "throwExceptionIfCannotOpen" in the openEnvironment()?

Comment (jakesmith) [reply esp/services/ws_ecl/ws_ecl_service.cpp]: > if the openEnvironment() returns nullptr, a default value is used. If the openEnvironment() throws exception, should we catch the exception and use the default value?

No, openEnvironment() should never return null and it's impossible for the current implementation  to do so, so the 'if (env)' check inside envGetConfigurationDirectory().
As long as openEnvironment is guaranteed to return an IConstEnvironment OR throw an exception, no check after each call to openEnvironment() is necessary.

Comment (jakesmith) [reply esp/services/ws_ecl/ws_ecl_service.cpp]: @wangkx - back to you.

**Comment by wangkx** [comment]
@ghalliday / @jakesmith I just made the changes based on your last review. Please review again. I pass 'true' into getEnvironmentFactory() except for: 1. get the factory for CSdsSubscription; 2. get the factory from local environment file in WsDeploy; 3. get the factory in configenv and envgen. Please let me know if that is incorrect.

File: common/environment/environment.cpp [hunk 7]
```diff
@@ -362,7 +362,7 @@ class CSdsSubscription : implements ISDSSubscription, public CInterface
     CSdsSubscription()
     {
         m_constEnvUpdated = false;
-        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory();
+        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(false);
```
Comment (jakesmith) [inline common/environment/environment.cpp]: not new but a bit odd that that this class (CSdsSubscription) - which is only ever created by CEnvironmentFactory, asks for the factory here.
It would be more obvious if the factory was passed into the CSdsSubscription constructor.

File: esp/services/WsDeploy/WsDeployService.hpp [hunk 8]
```diff
@@ -512,7 +512,7 @@ class CWsDeployFileInfo : public CInterface, implements IConfigFileObserver
         m_pGraphXml.clear();
         m_pNavTree.clear();
 
-        Owned<IEnvironmentFactory> factory = getEnvironmentFactory();       
+        Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
```
Comment (jakesmith) [inline esp/services/WsDeploy/WsDeployService.hpp]: There appears to be an independent subscription / update service in WsDeployService, but I'm not sure why ?
i.e. CWsDeployFileInfo::CSdsSubscription, which ends up calling environmentUpdated() , but the built in validateCache mechanism will be doing the same afaics...

**Comment by jakesmith** [comment]
>1. get the factory for CSdsSubscription;

Yes, though it's v. odd that it is calling get factory when it is basically a private class that the factory has created. It would make more sense and look cleaner if it passed in the factory into the constructor and used it directly.

> 2. get the factory from local environment file in WsDeploy;
> 3. get the factory in configenv and envgen. Please let me know if that is incorrect.

Yes, I think right not to be subscribing in these cases.

@wangkx

File: common/environment/environment.cpp (hunk 7 above)
Comment (wangkx) [reply common/environment/environment.cpp]: If I pass the factory 'envFactory' into the CSdsSubscription constructor and call envFactory->subscribe() directly inside the constructor , the factory CEnvironmentFactory has to be fully declared before the CSdsSubscription. I may have to add a new .h file and move many class declarations from environment.cpp to the .h file. @jakesmith do you want this big change?

File: esp/services/WsDeploy/WsDeployService.hpp (hunk 8 above)
Comment (wangkx) [reply esp/services/WsDeploy/WsDeployService.hpp]: I saw the following code inside the environmentUpdated(), Could that be why the independent subscription / update service is needed in WsDeployService?
'''
        m_pEnvXml.clear();
        m_pGraphXml.clear();
        m_pNavTree.clear();
'''

File: common/environment/environment.cpp (hunk 7 above)
Comment (jakesmith) [reply common/environment/environment.cpp]: > the factory CEnvironmentFactory has to be fully declared before the CSdsSubscription.

It does not need a CEnvironmentFactory, it is dealing with a IEnvironmentFactory, so it does not need to know a CEnvironmentFactory.

CEnvironmentFactory::validateCache() could be:
```
    virtual void validateCache()
    {
        if (!subscription)
            subscription.setown( new CSdsSubscription(*this) );

        subscription->handleEnvironmentChange();
    }
```
..and the CSdsSubscription constructor could be:
```
    CSdsSubscription(IEnvironmentFactory &_factory) : factory(_factory)
```

File: esp/services/WsDeploy/WsDeployService.hpp (hunk 8 above)
Comment (jakesmith) [reply esp/services/WsDeploy/WsDeployService.hpp]: I guess so, shame it's duplicating the mechanism though.
But it means it should not be calling getEnvironmentFactory with true, as this is installing it's own update mechanism.

**Comment by jakesmith** [comment]
@wangkx - added a couple of replies

**Comment by wangkx** [comment]
@jakesmith please review.

File: common/environment/environment.cpp [hunk 9]
```diff
@@ -359,10 +359,9 @@ void CLockedEnvironment::rollback()
 class CSdsSubscription : implements ISDSSubscription, public CInterface
 {
 public:
-    CSdsSubscription()
+    CSdsSubscription(IEnvironmentFactory* envFactory)
     {
         m_constEnvUpdated = false;
-        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(false);
```
Comment (jakesmith) [inline common/environment/environment.cpp]: @wangkx - what about other calls to getEnvironmentFactory in this class?

**Comment by wangkx** [comment]
@jakesmith sorry, my mistake. Please check my fix.

**Comment by jakesmith** [comment]
@wangkx - thanks. Can you now squash the commits ready for merge.

**Comment by wangkx** [comment]
@AttilaVamos could you please check the last Smoketest: HPCC Stop: Fail 1 components are still up!

**Comment by AttilaVamos** [comment]
@wangkx  Yes, I'm checking.

**Comment by wangkx** [comment]
Thanks!

**Comment by AttilaVamos** [comment]
@wangkx There was a leftover process from my previous test run (independent from your PR) and that caused problem at the end of your previous test. Now everything is fine.

**Comment by wangkx** [comment]
@jakesmith the commits have been squashed.

**Comment by jakesmith** [comment]
@richardkchapman - ready to merge.
